node/src/components/block_proposer.rs

```diff
@@ -380,33 +382,74 @@ impl BlockProposerReady {
     ) -> ProtoBlock {
         let max_transfers = deploy_config.block_max_transfer_count as usize;
         let max_deploys = deploy_config.block_max_deploy_count as usize;
+        let max_block_size_bytes = deploy_config.max_block_size as usize;
+        let block_gas_limit = Gas::from(deploy_config.block_gas_limit);
 
         let mut transfers = Vec::new();
         let mut wasm_deploys = Vec::new();
+        let mut block_gas_running_total = Gas::zero();
+        let mut block_size_running_total = 0usize;
 
         for (hash, deploy_type) in self.sets.pending.iter() {
-            let under_max_transfers = transfers.len() < max_transfers;
-            let under_max_deploys = wasm_deploys.len() < max_deploys;
-            if !under_max_deploys && !under_max_transfers {
+            let at_max_transfers = transfers.len() == max_transfers;
+            let at_max_deploys = wasm_deploys.len() == max_deploys;
+
+            // Early exit if block limits are met.
+            if block_gas_running_total >= block_gas_limit
+                && block_size_running_total >= max_block_size_bytes
+                && at_max_deploys
+                && at_max_transfers
```

Comment:
This should be a disjunction:
```rust
            if block_gas_running_total >= block_gas_limit
                || block_size_running_total >= max_block_size_bytes
                || (at_max_deploys && at_max_transfers)
```

Reply:
But I wonder whether we should somehow limit the number of iterations where we don't add a deploy instead. Otherwise if we're one byte below the size limit, we will iterate through all the rest of `pending` and never add anything.

At least we could already `break` if:
```rust
block_gas_running_total + min_gas_per_deploy > block_gas_limit
|| block_size_running_total + min_size_per_deploy > max_block_size_bytes
```
(…or rather also with `checked_add`.)

Reply:
For the time being, I think we can't exit early based on `block_gas_limit` because (and this reminds me to mention it to @mpapierski :) ) transfers currently have a payment amount of 0, until this constant has been defined. We would exit early before checking for more (in this case zero-cost) transfers.

Reply:
@dwerner I'm working on adding gas cost to wasmless transfers as part of https://casperlabs.atlassian.net/browse/EE-1122. That cost will come from EE's protocol data rather than being a constant defined somewhere. Would it be an issue if you'd have to query EE's protocol data to get wasmless transfer price here?

Reply:
@mpapierski what would be the lifetime of that value? I'm going to guess that it shouldn't need to be queried very frequently... but I could be wrong, of course.

Reply:
> 
> @mpapierski what would be the lifetime of that value? I'm going to guess that it shouldn't need to be queried very frequently... but I could be wrong, of course.

That's right. This value won't change for given protocol version